### PR TITLE
Move TLS related code from test_suite_ssl.function to ssl_helpers.c and ssl_helpers.h

### PR DESCRIPTION
**Pull request by zhangsenWang**
## Description
`test_suite_ssl.function` contains many functions that are used to set up a TLS connection. Move those functions (and the types they use) to a new file `tests/src/ssl_helpers.c` with declarations in `tests/include/test/ssl_helpers.h`

## Status
**READY**

## Requires Backporting
Yes, 2.28.

## Additional comments
target issue #6173 

## Todos
- [x] create `ssl_helpers.h` and `ssl_helpers.c`
- [x] rename types that need to be moved
- [x] move types to `ssl_helpers.h`
- [x] rename functions that need to be moved
- [x] adjust format to limit line length 
- [x] move functions to `ssl_helpers.c` and add declaration to `ssl_helpers.h`
- [ ] backport

### CODE REVIEW

File: tests/src/ssl_helpers.c [hunk 1]
```diff
@@ -0,0 +1,2205 @@
...
+}
+#endif /* MBEDTLS_TEST_HOOKS */
+
+void mbedtls_test_ssl_buffer_init( mbedtls_test_ssl_buffer *buf )
+{
+    memset( buf, 0, sizeof( *buf ) );
+}
+
+int mbedtls_test_ssl_buffer_setup( mbedtls_test_ssl_buffer *buf,
+                                                            size_t capacity )
```
Comment (xkqian) [inline tests/src/ssl_helpers.c:119]:
```suggestion
                                size_t capacity )
```
Vertical alligned

File: tests/src/ssl_helpers.c [hunk 2]
```diff
@@ -0,0 +1,2205 @@
...
+void mbedtls_test_ssl_buffer_free( mbedtls_test_ssl_buffer *buf )
+{
+    if( buf->buffer != NULL )
+        mbedtls_free( buf->buffer );
+
+    memset( buf, 0, sizeof( *buf ) );
+}
+
+int mbedtls_test_ssl_buffer_put( mbedtls_test_ssl_buffer *buf,
+                             const unsigned char *input, size_t input_len )
```
Comment (xkqian) [inline tests/src/ssl_helpers.c:139]:
```suggestion
                                 const unsigned char *input, size_t input_len )
```
Vertical alligned

File: tests/src/ssl_helpers.c [hunk 3]
```diff
@@ -0,0 +1,2205 @@
...
+        memcpy( buf->buffer + buf->start + buf->content_length - buf->capacity,
+                input, input_len );
+    }
+
+    buf->content_length += input_len;
+    return input_len;
+}
+
+int mbedtls_test_ssl_buffer_get( mbedtls_test_ssl_buffer *buf,
+                             unsigned char* output, size_t output_len )
```
Comment (xkqian) [inline tests/src/ssl_helpers.c:188]: vertical alligned

File: tests/src/ssl_helpers.c [hunk 4]
```diff
@@ -0,0 +1,2205 @@
...
+
+    memset( buf, 0, sizeof( *buf ) );
+}
+
+int mbedtls_test_ssl_buffer_put( mbedtls_test_ssl_buffer *buf,
+                             const unsigned char *input, size_t input_len )
+{
+    size_t overflow = 0;
+
+    if( ( buf == NULL ) || ( buf->buffer == NULL ) )
```
Comment (xkqian) [inline tests/src/ssl_helpers.c:143]:
```suggestion
    if( buf == NULL || buf->buffer == NULL )
```

File: tests/src/ssl_helpers.c [hunk 5]
```diff
@@ -0,0 +1,2205 @@
...
+        return -1;
+
+    /* Reduce input_len to a number that fits in the buffer. */
+    if ( ( buf->content_length + input_len ) > buf->capacity )
```
Comment (xkqian) [inline tests/src/ssl_helpers.c:147]:
```suggestion
    if ( buf->content_length + input_len > buf->capacity )
```

File: tests/src/ssl_helpers.c [hunk 6]
```diff
@@ -0,0 +1,2205 @@
...
+    buf->content_length += input_len;
+    return input_len;
+}
+
+int mbedtls_test_ssl_buffer_get( mbedtls_test_ssl_buffer *buf,
+                             unsigned char* output, size_t output_len )
+{
+    size_t overflow = 0;
+
+    if( ( buf == NULL ) || ( buf->buffer == NULL ) )
```
Comment (xkqian) [inline tests/src/ssl_helpers.c:192]:
```suggestion
    if( buf == NULL || buf->buffer == NULL )
```

File: tests/src/ssl_helpers.c [hunk 7]
```diff
@@ -0,0 +1,2205 @@
...
+    if( output == NULL && output_len == 0 )
+        return 0;
+
+    if( buf->content_length < output_len )
+        output_len = buf->content_length;
+
+    /* Calculate the number of bytes that need to be drawn from lower memory
+     * address */
+    if( buf->start + output_len > buf->capacity )
+    {
```
Comment (xkqian) [inline tests/src/ssl_helpers.c:204]:
```suggestion
```

File: tests/src/ssl_helpers.c [hunk 8]
```diff
@@ -0,0 +1,2205 @@
...
+        overflow = ( buf->start + output_len ) % buf->capacity;
+    }
```
Comment (xkqian) [inline tests/src/ssl_helpers.c:206]:
```suggestion
```

File: tests/src/ssl_helpers.c [hunk 9]
```diff
@@ -0,0 +1,2205 @@
...
+
+    buf->content_length -= output_len;
+    buf->start = ( buf->start + output_len ) % buf->capacity;
+
+    return output_len;
+}
+
+int mbedtls_test_ssl_message_queue_setup( mbedtls_test_ssl_message_queue *queue,
+                                      size_t capacity )
```
Comment (xkqian) [inline tests/src/ssl_helpers.c:221]: vertical alligned

File: tests/src/ssl_helpers.c [hunk 10]
```diff
@@ -0,0 +1,2205 @@
...
+    queue->pos %= queue->capacity;
+    if( queue->pos < 0 )
+        queue->pos += queue->capacity;
+
+    return ( message_length > buf_len ) ? buf_len : message_length;
+}
+
+int mbedtls_test_ssl_message_queue_peek_info(
+                                        mbedtls_test_ssl_message_queue *queue,
+                                        size_t buf_len, size_t* msg_len )
```
Comment (xkqian) [inline tests/src/ssl_helpers.c:285]:
```suggestion
                                        size_t buf_len, size_t *msg_len )
```

File: tests/src/ssl_helpers.c [hunk 11]
```diff
@@ -0,0 +1,2205 @@
...
+    *msg_len = queue->messages[queue->pos];
+    return ( *msg_len > buf_len ) ? MBEDTLS_TEST_ERROR_MESSAGE_TRUNCATED : 0;
+}
+
+void mbedtls_test_mock_socket_init( mbedtls_test_mock_socket *socket )
+{
+    memset( socket, 0, sizeof( *socket ) );
+}
+
+void mbedtls_test_mock_socket_close( mbedtls_test_mock_socket* socket )
```
Comment (xkqian) [inline tests/src/ssl_helpers.c:301]:
```suggestion
void mbedtls_test_mock_socket_close( mbedtls_test_mock_socket *socket )
```

File: tests/src/ssl_helpers.c [hunk 12]
```diff
@@ -0,0 +1,2205 @@
...
+    }
+
+    if( socket->peer != NULL )
+        memset( socket->peer, 0, sizeof( *socket->peer ) );
+
+    memset( socket, 0, sizeof( *socket ) );
+}
+
+int mbedtls_test_mock_socket_connect( mbedtls_test_mock_socket* peer1,
+                                 mbedtls_test_mock_socket* peer2,
```
Comment (xkqian) [inline tests/src/ssl_helpers.c:325]: vertical alligned

File: tests/src/ssl_helpers.c [hunk 13]
```diff
@@ -0,0 +1,2205 @@
...
+        (mbedtls_test_ssl_buffer*) mbedtls_calloc( 1,
+                                            sizeof(mbedtls_test_ssl_buffer) );
+    if( peer1->output == NULL )
+    {
+        ret = MBEDTLS_ERR_SSL_ALLOC_FAILED;
+        goto exit;
+    }
+    mbedtls_test_ssl_buffer_init( peer1->output );
+    if( 0 != ( ret = mbedtls_test_ssl_buffer_setup( peer1->output, bufsize ) ) )
+    {
```
Comment (xkqian) [inline tests/src/ssl_helpers.c:340]:
```suggestion
```

File: tests/src/ssl_helpers.c [hunk 14]
```diff
@@ -0,0 +1,2205 @@
...
+        goto exit;
+    }
```
Comment (xkqian) [inline tests/src/ssl_helpers.c:342]:
```suggestion
```

File: tests/src/ssl_helpers.c [hunk 15]
```diff
@@ -0,0 +1,2205 @@
...
+        (mbedtls_test_ssl_buffer*) mbedtls_calloc( 1,
+                                            sizeof(mbedtls_test_ssl_buffer) );
+    if( peer2->output == NULL )
+    {
+        ret = MBEDTLS_ERR_SSL_ALLOC_FAILED;
+        goto exit;
+    }
+    mbedtls_test_ssl_buffer_init( peer2->output );
+    if( 0 != ( ret = mbedtls_test_ssl_buffer_setup( peer2->output, bufsize ) ) )
+    {
```
Comment (xkqian) [inline tests/src/ssl_helpers.c:354]:
```suggestion
```

File: tests/src/ssl_helpers.c [hunk 16]
```diff
@@ -0,0 +1,2205 @@
...
+        goto exit;
+    }
```
Comment (xkqian) [inline tests/src/ssl_helpers.c:356]:
```suggestion
```

File: tests/src/ssl_helpers.c [hunk 17]
```diff
@@ -0,0 +1,2205 @@
...
+
+void mbedtls_test_message_socket_init( mbedtls_test_message_socket_context *ctx )
+{
+    ctx->queue_input = NULL;
+    ctx->queue_output = NULL;
+    ctx->socket = NULL;
+}
+
+int mbedtls_test_message_socket_setup( mbedtls_test_ssl_message_queue* queue_input,
+                                  mbedtls_test_ssl_message_queue* queue_output,
```
Comment (xkqian) [inline tests/src/ssl_helpers.c:435]: vertical alligned

File: tests/src/ssl_helpers.c [hunk 18]
```diff
@@ -0,0 +1,2205 @@
...
+    if( ctx == NULL )
+        return;
+
+    mbedtls_test_ssl_message_queue_free( ctx->queue_input );
+    mbedtls_test_mock_socket_close( ctx->socket );
+    memset( ctx, 0, sizeof( *ctx ) );
+}
+
+int mbedtls_test_mock_tcp_send_msg( void *ctx, const unsigned char *buf,
+                                                                    size_t len )
```
Comment (xkqian) [inline tests/src/ssl_helpers.c:462]: vertical alligned

File: tests/src/ssl_helpers.c [hunk 19]
```diff
@@ -0,0 +1,2205 @@
...
+{
+    mbedtls_test_ssl_message_queue* queue;
+    mbedtls_test_mock_socket* socket;
+    mbedtls_test_message_socket_context *context =
+                                    (mbedtls_test_message_socket_context*) ctx;
+
+    if( context == NULL || context->socket == NULL
+        || context->queue_output == NULL )
+    {
```
Comment (xkqian) [inline tests/src/ssl_helpers.c:471]:
```suggestion
```

File: tests/src/ssl_helpers.c [hunk 20]
```diff
@@ -0,0 +1,2205 @@
...
+        return MBEDTLS_TEST_ERROR_CONTEXT_ERROR;
+    }
```
Comment (xkqian) [inline tests/src/ssl_helpers.c:473]:
```suggestion
```

File: tests/src/ssl_helpers.c [hunk 21]
```diff
@@ -0,0 +1,2205 @@
...
+            psa_destroy_key( key_slot );
+#endif
+        mbedtls_pk_free( &( cert->pkey ) );
+    }
+
+    return ret;
+}
+
+int mbedtls_test_ssl_endpoint_init( mbedtls_test_ssl_endpoint *ep,
+                            int endpoint_type,
```
Comment (xkqian) [inline tests/src/ssl_helpers.c:680]: vertical alligned

File: tests/src/ssl_helpers.c [hunk 22]
```diff
@@ -0,0 +1,2205 @@
...
+    ret = mbedtls_ctr_drbg_seed( &( ep->ctr_drbg ), mbedtls_entropy_func,
+                    &( ep->entropy ), (const unsigned char *) ( ep->name ),
+                    strlen( ep->name ) );
+    TEST_ASSERT( ret == 0 );
+
+    /* Non-blocking callbacks without timeout */
+    if( dtls_context != NULL )
+    {
+        mbedtls_ssl_set_bio( &( ep->ssl ), dtls_context,
+            mbedtls_test_mock_tcp_send_msg,
```
Comment (xkqian) [inline tests/src/ssl_helpers.c:739]: vertical alligned

File: tests/src/ssl_helpers.c [hunk 23]
```diff
@@ -0,0 +1,2205 @@
...
+            mbedtls_test_mock_tcp_recv_msg,
+            NULL );
+    }
+    else
+    {
+        mbedtls_ssl_set_bio( &( ep->ssl ), &( ep->socket ),
+            mbedtls_test_mock_tcp_send_nb,
```
Comment (xkqian) [inline tests/src/ssl_helpers.c:746]: ditto

File: tests/src/ssl_helpers.c [hunk 24]
```diff
@@ -0,0 +1,2205 @@
...
+        mbedtls_svc_key_id_t *key_slot = cert->pkey.pk_ctx;
+
+        psa_destroy_key( *key_slot );
+    }
+#endif
+    mbedtls_pk_free( &( cert->pkey ) );
+}
+
+void mbedtls_test_ssl_endpoint_free( mbedtls_test_ssl_endpoint *ep,
+                            mbedtls_test_message_socket_context *context )
```
Comment (xkqian) [inline tests/src/ssl_helpers.c:812]: ditto

File: tests/src/ssl_helpers.c [hunk 25]
```diff
@@ -0,0 +1,2205 @@
...
+{
+    mbedtls_test_ssl_endpoint_certificate_free( ep );
+
+    mbedtls_ssl_free( &( ep->ssl ) );
+    mbedtls_ssl_config_free( &( ep->conf ) );
+    mbedtls_ctr_drbg_free( &( ep->ctr_drbg ) );
+    mbedtls_entropy_free( &( ep->entropy ) );
+
+    if( context != NULL )
+    {
```
Comment (xkqian) [inline tests/src/ssl_helpers.c:822]: Remove the '{' and '}' , the same as line 824,826,828

File: tests/src/ssl_helpers.c [hunk 26]
```diff
@@ -0,0 +1,2205 @@
...
+        mbedtls_test_message_socket_close( context );
+    }
+    else
+    {
+        mbedtls_test_mock_socket_close( &( ep->socket ) );
+    }
+}
+
+int mbedtls_test_move_handshake_to_state( mbedtls_ssl_context *ssl,
+                                     mbedtls_ssl_context *second_ssl,
```
Comment (xkqian) [inline tests/src/ssl_helpers.c:832]: ditto

File: tests/src/ssl_helpers.c [hunk 27]
```diff
@@ -0,0 +1,2205 @@
...
+
+#endif /* MBEDTLS_X509_CRT_PARSE_C && MBEDTLS_ENTROPY_C && MBEDTLS_CTR_DRBG_C */
+
+int mbedtls_ssl_write_fragment( mbedtls_ssl_context *ssl, unsigned char *buf,
+                                int buf_len, int *written,
+                                const int expected_fragments )
+{
+    int ret = mbedtls_ssl_write( ssl, buf + *written, buf_len - *written );
+    if( ret > 0 )
+    {
```
Comment (xkqian) [inline tests/src/ssl_helpers.c:880]: ditto

File: tests/src/ssl_helpers.c [hunk 28]
```diff
@@ -0,0 +1,2205 @@
...
+
+    return 0;
+
+exit:
+    /* Some of the tests failed */
+    return -1;
+}
+
+void set_ciphersuite( mbedtls_ssl_config *conf, const char *cipher,
+                      int* forced_ciphersuite )
```
Comment (xkqian) [inline tests/src/ssl_helpers.c:952]:
```suggestion
                      int *forced_ciphersuite )
```

File: tests/src/ssl_helpers.c [hunk 29]
```diff
@@ -0,0 +1,2205 @@
...
+    }
+
+    mbedtls_ssl_conf_ciphersuites( conf, forced_ciphersuite );
+
+exit:
+    return;
+}
+
+int psk_dummy_callback( void *p_info, mbedtls_ssl_context *ssl,
+                  const unsigned char *name, size_t name_len )
```
Comment (xkqian) [inline tests/src/ssl_helpers.c:981]: ditto

File: tests/src/ssl_helpers.c [hunk 30]
```diff
@@ -0,0 +1,2205 @@
...
+    (void) name;
+    (void) name_len;
+
+    return ( 0 );
+}
+
+#if defined(MBEDTLS_SSL_PROTO_TLS1_2) && \
+    defined(MBEDTLS_CIPHER_MODE_CBC) && defined(MBEDTLS_AES_C)
+int mbedtls_test_psa_cipher_encrypt_helper( mbedtls_ssl_transform *transform,
+                                        const unsigned char *iv, size_t iv_len,
```
Comment (xkqian) [inline tests/src/ssl_helpers.c:994]: ditto

File: tests/src/ssl_helpers.c [hunk 31]
```diff
@@ -0,0 +1,2205 @@
...
+    return( 0 );
+#else
+    return mbedtls_cipher_crypt( &transform->cipher_ctx_enc,
+                            iv, iv_len, input, ilen, output, olen );
+#endif /* MBEDTLS_USE_PSA_CRYPTO */
+}
+#endif /* MBEDTLS_SSL_PROTO_TLS1_2 && MBEDTLS_CIPHER_MODE_CBC && MBEDTLS_AES_C */
+
+int mbedtls_test_ssl_build_transforms( mbedtls_ssl_transform *t_in,
+                                    mbedtls_ssl_transform *t_out,
```
Comment (xkqian) [inline tests/src/ssl_helpers.c:1036]: vertical alligned

**Review by xkqian — CHANGES_REQUESTED** [review]
I haven't gone through all of the code, just give out draft comments.   Please pay attention to the line width < 80 characters and vertical alligned.

**Comment by zhangsenWang** [comment]
The line width and vertical align issues have be revised.  Besides, in the process of finding out reason of some check failure, I encounter some points and finally solve them in my own way. I'm not sure if it's suitable.

1. I add one library `mbedtls_ssl_test` in `CMakeLists.txt`. The reason is that since all files in `tests/src` are compiled into an OBJECT library `mbedtls_test`, but our `tests/src/ssl_helpers.c` need many functions in the `library` , other programs (such as those in `programs`) that need library `mbedtls_test `will have to link the functions in the `library`. which will cause undefined reference.  This error only occurs when using cmake, nothing need to be changed with the `Makefile`

2. use `#pragma` in `tests/src/ssl_helpers.c` to disable a warning `C4267 `about conversion from `size_t `to `int `etc. This warning is treated as error. This is because original `tests/test_suite_ssl.c` is compile with `tests/CMakeLists.txt` but now `tests/src/ssl_helpers.c` is compiled with `CMakeLists.txt`. The `CMAKE_C_CFLAGS `in these two file are not the same. This only   occurs under MSVC

I'm looking forward for your opinions, thanks.  @xkqian @gilles-peskine-arm

**Comment by davidhorstmann-arm** [comment]
Was this superseded by #6500? Can it be closed?

**Comment by gilles-peskine-arm** [comment]
Superseded by https://github.com/Mbed-TLS/mbedtls/pull/6500

**Pull request closed without merging**